pkg/apis/core/types_cloudprofile.go

```diff
@@ -78,8 +78,20 @@ type KubernetesSettings struct {
 type MachineImage struct {
 	// Name is the name of the image.
 	Name string
-	// Versions contains versions and expiration dates of the machine image
-	Versions []ExpirableVersion
+	// Versions contains versions, expiration dates and container runtimes of the machine image
+	Versions []MachineImageVersion
+}
+
+// MachineImageVersion is an expirable version with list of supported container runtimes and interfaces
+type MachineImageVersion struct {
+	// Version is the version identifier.
+	Version string
+	// ExpirationDate defines the time at which this version expires.
+	ExpirationDate *metav1.Time
+	// ContainerRuntimeInterfaces is list of container runtime interface names that are available for this image version.
+	ContainerRuntimeInterfaces []string
```

Comment:
I am not sure if we should put the Container Runtime and CRI information into each version - this will mean a lot of duplicate information when the CRI and runtime list grows. Maybe we can find a better way - e.g on the  a dedicated CRI section with reference to the images.

Not every container runtime interface can work with every container runtime (e.g CRI-O cannot work with gvisor et al.). I do not see a way to express that here. I think thats another reason for a dedicated CRI section instead.

Also I think it is kind of nice to have the expirable versions for machine images and kubernetes versions. This is also being taken advantage of when implementing the [Version Management as part of GEP5](https://github.com/gardener/gardener/pull/2107). 

```yaml
  cri:
    - name: containerd
      machineImage:
        - name: ubuntu
          versionConstraint: ">=16.04.0"
      containerRuntimes:
        - gvisor
        - kata
    - name: cri-o
      machineImage:
        - name: ubuntu
          versions:
            - "18.04.0"
            - "19.04.0"
            - "19.10.0"
      containerRuntimes:
        - podman
`

Regarding the example above: gVisor for instance needs Linux Kernel version > 4.14.77+, and Ubuntu 16.04 that is required for containerd is already [based on kernel 4.4](https://wiki.ubuntu.com/XenialXerus/ReleaseNotes). It seems to me that when containerd is supported, then the container runtimes are also supported -> no need to separately restrict the container runtimes to the os.  I would have to check that for others too.

Reply:
and that would also address my comment to cross-check CRI with Container runtimes. One point though, some runtimes (e.g., gvisor) would require a certain version of the OS, and although containerd is right on there, the os version required by containerd can be less than the one required by gvisor and so it can lead to unexpected side-effects, i.e., you cant assume that gvisor is supported as long as containerd is on the system.

Reply:
Yes I think that would make sense. This way we are not overcomplicating it. Also we only miss out on some small portion of the versions.

Reply:
First of all CRI is maintained in-tree by the gardener, https://github.com/gardener/gardener/blob/82af7cd270c092cd07a32455d86c38643a299e74/charts/seed-operatingsystemconfig/original/templates/kubelet/_kubelet.service#L9-L16

And nowadays the CRI can be either omitted or a list with a single element `containerd`. And container runtimes rely on `containerd` following the contract in the [GEP](https://github.com/gardener/gardener/blob/master/docs/proposals/10-shoot-additional-container-runtimes.md)

> The first phase would be to allow the change of the Kubelet configuration as described here so that Kubernetes would use containerd instead of the default Dockershim

So the contract in the GEP implies that every container runtime, except docker, must be working with `containerd`. 

I was thinking of a list of unsupported CR(I)s and started the implementation in this way, but then the operator is responsible to explicitly disable broken CR(I)s for particular image and version. Thus I switched to list of enabled CR(I)s. 



> I am not sure if we should put the Container Runtime and CRI information into each version - this will mean a lot of duplicate information when the CRI and runtime list grows.

We can make CRI configuration also on image level if it is applicable to all versions. What do you think?

> Maybe we can find a better way - e.g on the a dedicated CRI section with reference to the images.

How does this help to keep the list short? Yes, semver version restrictions can be used, but this does not guarantee a shorter list.  The cloud profile is a resource that maintain all constraints for the environment, thus I prefer it to be as explicit as possible. 

> Also I think it is kind of nice to have the expirable versions for machine images and kubernetes versions. This is also being taken advantage of when implementing the Version Management as part of GEP5.

I do not understand, every image version still have its own `expirationDate` field so what do you mean?

Reply:
> And container runtimes rely on containerd following the contract in the GEP

Yes, but containerd is not the only CRI available, maybe we will add for instance CRI-O in the future or other ones. I think that below can be interpreted  that crio-o can use gVisor as the container runtime, but only the combination containerd -> gvisor && containerd-> kata are allowed. :
```yaml 
    containerRuntimeInterfaces:
     - containerd
     - cri-o
     containerRuntimes:
     - gvisor
     - kata-containers
`

> How does this help to keep the list short? Yes, semver version restrictions can be used, but this does not guarantee a shorter list. The cloud profile is a resource that maintain all constraints for the environment, thus I prefer it to be as explicit as possible.

Because we do not repeat the information which CRI and container runtime is support for each and every version.For instance containerd and gvisor can work with any ubuntu if the linux kernel version is > 4.14.77. So we do not have to repeat the same information that containerd is supported every time.
Rather we can express that with a semVer constraint. 
If that is not applicable, e.g crio-o and ubuntu (see above) - we can still specify single versions. 

> I do not understand, every image version still have its own expirationDate field so what do you mean?

GEP5 simplifies the handling of machine image versions and kubernetes versions by using common functionality for expirableVersions. If machine images do not use expirableVersions any more, that cannot be done / or needs a hacky conversion.

Reply:
> I think that below can be interpreted that crio-o can use gVisor as the container runtime, but only the combination containerd -> gvisor && containerd-> kata are allowed. 

Generally you are right, but `gvisor` and `kata` extensions can implement a validation webhook to check whether they are configured with proper CRI.

I am afraid that if we move CRI configs to dedicated field, we will more or less repeat the image names and versions there. So we are just moving the duplicated information.

Reply:
I feel like the approach with Webhooks is too cumbersome - we do not need that with the above mentioned approach.

> I am afraid that if we move CRI configs to dedicated field, we will more or less repeat the image names and versions there. So we are just moving the duplicated information.

Some redundancy is not avoidable here I think. But I do not see why moving the CRI information like describe in my previous comment increases redundancy, it reduces it compared to moving it to the image or the machine image version.

I feel like we need another opinion here.

Reply:
Valid points, thanks for bringing up the doubts. After thinking about it I tend to agree with @vpnachev. The proposal of @danielfoehrKn looks like a lot of double maintenance wrt machine images (it's better/easier-to-maintain to store the CRI/container runtime information directly where a machine image version is defined).

Though, while I understand the webhook validation suggestion, I think for this use-case it's not perfectly applicable. From end-user's pov you don't see a potential limitation (e.g., if gvisor cannot be used with cri-o) - but that's one important use-case of the `CloudProfile`.

So, what about this:

```yaml
spec:
  machineImages:
  - name ubuntu:
    versions:
    - version: 1.2.3
      cri:
      - name: containerd
        containerRuntimes:
        - gvisor
        - kata
      - name: cri-o
        containerRuntimes:
        - kata
``` 

> Also I think it is kind of nice to have the expirable versions for machine images and kubernetes versions. This is also being taken advantage of when implementing the Version Management as part of GEP5.

If @vpnachev embeds the `core(/{v1alpha,v1beta1}).ExpireableVersion` type into the `core(/{v1alpha,v1beta1}).MachineImageVersion` type then you, @danielfoehrKn, should still be able to work with the one `ExpireableVersion` type used in both Kubernetes and machine image versions for your GEP5 PRs.

Reply:
Alright, so we'll have more duplicate information, but less maintenance. When adding a version,  we do not have to adjust another portion of the ClouProfile, but instead can just add the version. Fine with me.

> @danielfoehrKn, should still be able to work with the one ExpireableVersion type used in both Kubernetes and machine image versions for your GEP5 PRs.

Good suggestion. That would significantly decrease the needed refactoring.